### PR TITLE
lightning: support compression when sending kv pairs to tikv

### PR DESCRIPTION
**Pull request by gozssky**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/41163

Problem Summary:
Support compression when sending kv pairs to tikv.

### What is changed and how it works?
Add a new config item `compress-kv-pairs` under [tikv-importer] to control the behavior of whether to compress the kv pairs before sending them to tikv. User can set it to `gzip` to enable compression. Default is disabled.

In integration test `lightning_import_compress`, the compression ratio can be as high as 10x. 30700184B (uncompress) vs 2992553B (gzip).

### Check List

Tests 

- [x] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
lightning: support compression when sending kv pairs to tikv
```

### CODE REVIEW

File: br/tests/lightning_import_compress/run.sh [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+uncompress=$(grep "import write" /tmp/backup_restore_test/lightning-import-compress1.log |
+  grep -Eo "bytes=[0-9]+" | sed 's/bytes=//g' | awk '{sum+=$1} END {print sum}')
+gzip=$(grep "import write" /tmp/backup_restore_test/lightning-import-compress2.log |
+  grep -Eo "bytes=[0-9]+" | sed 's/bytes=//g' | awk '{sum+=$1} END {print sum}')
+gz=$(grep "import write" /tmp/backup_restore_test/lightning-import-compress3.log |
+  grep -Eo "bytes=[0-9]+" | sed 's/bytes=//g' | awk '{sum+=$1} END {print sum}')
+
+echo "uncompress: ${uncompress}, gzip: ${gzip}, gz: ${gz}"
+if [ "$uncompress" -lt "$gzip" ] || [ "$uncompress" -lt "$gz" ]; then
```
Comment (gozssky) [inline br/tests/lightning_import_compress/run.sh:53]:
```suggestion
if [ "$uncompress" -le "$gzip" ] || [ "$uncompress" -le "$gz" ]; then
```

**Review by dsdashun — APPROVED** [review]

File: br/pkg/lightning/config/config_test.go [hunk 2]
```diff
@@ -1144,3 +1144,17 @@ func TestCreateSeveralConfigsWithDifferentFilters(t *testing.T) {
...
+	require.Equal(t, config.CompressionTypeNone, ct)
+	require.NoError(t, ct.FromStringValue("gzip"))
+	require.Equal(t, config.CompressionTypeGzip, ct)
+	require.NoError(t, ct.FromStringValue("gz"))
+	require.Equal(t, config.CompressionTypeGzip, ct)
+	require.EqualError(t, ct.FromStringValue("zstd"), "invalid compression-type 'zstd', please choose valid option between ['gzip']")
+
+	require.Equal(t, "", config.CompressionTypeNone.String())
+	require.Equal(t, "gzip", config.CompressionTypeGzip.String())
+}
```
Comment (lance6716) [inline br/pkg/lightning/config/config_test.go:1160]: maybe add an unit test to make sure "gzip" is registered in grpc by `encoding.GetCompressor`

**Review by lance6716 — APPROVED** [review]

File: br/pkg/lightning/config/config_test.go (hunk 2 above)
Comment (gozssky) [reply br/pkg/lightning/config/config_test.go:1160]: Addressed in https://github.com/pingcap/tidb/pull/41164/commits/1c3c845c802a7930d6d960136f5ced66502eac6f. I add a `swtich` in `makeConn`. Now we don't reply on the string of compression type.

**Comment by gozssky** [comment]
Test case passed.
image

**Comment by gozssky** [comment]
CI almost passed.
https://ci.pingcap.net/blue/organizations/jenkins/br_ghpr_unit_and_integration_test/detail/br_ghpr_unit_and_integration_test/15052/pipeline/

**Comment by hawkingrei** [comment]
/hold

CI is too busy.

**Pull request merged**